### PR TITLE
Add additional support for array indexing in Set

### PR DESCRIPTION
**Pull request by JoshuaC215**
**Description**: Closes https://github.com/buger/jsonparser/issues/117. You can now use `Set` with array indices when the full path does not already exist, and supports `[+]` for append and `[-]` for prepend.

**Updated Logic:**
- if the array and numeric index exists, update it
- if the array exists and [+], append it
- if the array exists and [-], prepend it
- if the array exists and numeric index does not, pad with `null` as needed
- if the array does not exist and [0] or [+/-], create the array and set at 0
- if the array does not exist and non-zero numeric index, pad with `null` as needed

**Benchmark before change**:
```
BenchmarkJsonParserLarge-4                         50000            122002 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserMedium-4                       500000             23302 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserDeleteMedium-4                 500000             19750 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserEachKeyManualMedium-4         1000000             16974 ns/op             112 B/op          2 allocs/op
BenchmarkJsonParserEachKeyStructMedium-4          300000             22338 ns/op             672 B/op         12 allocs/op
BenchmarkJsonParserObjectEachStructMedium-4       200000             31673 ns/op             624 B/op         11 allocs/op
BenchmarkJsonParserSmall-4                       3000000              2227 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserEachKeyManualSmall-4          3000000              2255 ns/op              80 B/op          2 allocs/op
BenchmarkJsonParserEachKeyStructSmall-4          2000000              3258 ns/op             256 B/op          9 allocs/op
BenchmarkJsonParserObjectEachStructSmall-4       3000000              2926 ns/op             240 B/op          8 allocs/op
BenchmarkJsonParserSetSmall-4                    2000000              4072 ns/op             816 B/op          5 allocs/op
BenchmarkJsonParserDelSmall-4                    3000000              3447 ns/op               0 B/op          0 allocs/op
```

**Benchmark after change**:
```
BenchmarkJsonParserLarge-4                         50000            111986 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserMedium-4                       500000             21889 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserDeleteMedium-4                1000000             18926 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserEachKeyManualMedium-4          300000             20039 ns/op             112 B/op          2 allocs/op
BenchmarkJsonParserEachKeyStructMedium-4          500000             22492 ns/op             672 B/op         12 allocs/op
BenchmarkJsonParserObjectEachStructMedium-4       200000             31212 ns/op             624 B/op         11 allocs/op
BenchmarkJsonParserSmall-4                       5000000              2035 ns/op               0 B/op          0 allocs/op
BenchmarkJsonParserEachKeyManualSmall-4          3000000              2150 ns/op              80 B/op          2 allocs/op
BenchmarkJsonParserEachKeyStructSmall-4          2000000              3209 ns/op             256 B/op          9 allocs/op
BenchmarkJsonParserObjectEachStructSmall-4       3000000              2894 ns/op             240 B/op          8 allocs/op
BenchmarkJsonParserSetSmall-4                    2000000              4085 ns/op             816 B/op          5 allocs/op
BenchmarkJsonParserDelSmall-4                    2000000              3633 ns/op               0 B/op          0 allocs/op
```

### CODE REVIEW

**Comment by JoshuaC215** [comment]
See initial notes and discussion on https://github.com/qntfy/jsonparser/pull/1, where first commit code came from. Thanks @chilland!

**Comment by ryanleary** [comment]
The proposed logic in the description LGTM

**Comment by JoshuaC215** [comment]
Note on review: currently `Set()` returns `KeyPathNotFound` error if the top level json is an array. Only top level objects are supported. Not planning to add top level array support with this PR. If we wanted to add it we would do it here:
```
			// We can't set a top-level key if data isn't an object
			if len(data) == 0 || data[firstToken] != '{' {
				return nil, KeyPathNotFoundError
			}
```
Let me know if we'd like to add this to scope in the near term and I'm happy to file an issue / add it.

**Comment by ryanleary** [comment]
Is this no longer WIP then @JoshKCarroll ?

**Comment by JoshuaC215** [comment]
@ryanleary @chilland this is now complete and ready for review. Do we have ability to squash in github or do I need to squash manually? Also can we turn on https://travis-ci.org/qntfy/jsonparser ?

**Comment by ryanleary** [comment]
Travis enabled. We can squash and merge here.

File: parser.go [hunk 1]
```diff
@@ -555,27 +555,84 @@ var (
...
+			if idxNum >= 0 {
+				isArray = true
+				idx = idxNum
+			}
+		}
+	}
+	return isArray, idx
+}
+
+func createInsertComponent(keys []string, setValue []byte, startComma, endComma, isObject bool) []byte {
```
Comment (ryanleary) [inline parser.go]: Could you add a couple comments to this (i.e. a top-level "what does this function do?" comment) and single-line comments on the major blocks?

**Comment by ryanleary** [comment]
I *think* this looks good. A handful of clarifying comments will go a long way toward its understandability and maintainability I think.

**Comment by JoshuaC215** [comment]
@ryanleary does this sufficiently clarify? Also, any thought on versioning? Should we do any release before this PR? Feels a little weird to set this at 1.0.0.

**Comment by ryanleary** [comment]
Looks good to me. Thanks for adding the comments. 

No strong opinions on versioning really. Probably makes sense to divorce any version tags in this repo from the original one since they won't stay in sync anyway?

**Comment by JoshuaC215** [comment]
They never did any versioned releases so it's good there. We've just been versioning by commit.

**Comment by JoshuaC215** [comment]
Actually I don't have dev rights on this repo so you or @chilland will need to merge and tag :)

**Pull request merged**